[DRAFT] Implement DPP Reduction in wavefront

stefankoncarevic opened the pull request:
This pull request addresses the implementation of wavefront reduction using DPP (Data Parallel Primitives) instructions. Currently, the implementation covers case that there aren't sufficient threads for a parallel reduction, allowing each individual thread to perform its own reduction.

This is the link to ticket: https://github.com/ROCm/rocMLIR-internal/issues/1524

**Current Architecture Support**
The code currently supports the MI200 and MI300 architectures. For RDNA and MI100 architectures, Need to adjust the conditions based on the architecture and select the appropriate intrinsic instructions supported by these specific architectures.

**Considered Approaches**
_Kernel Pipeline Approach:_
Began by implementing an option for the chipset in the -rock-blockwise-gemm-to-threadwise pass. This option would be available for determining conditions to implement the desired functionality. It involves adding this option to the kernel pipeline, and modifying the our tools rocmlir-driver -c to pass the architecture information for generating the correct instructions for mlir passes. Feedback is needed to assess whether it's beneficial to add this option to the kernel pipeline pass.

_ROCDL Backend Approach:_
Since the chipset is already integrated within the ROCDL dialect backend pipeline, and instructions are lowered in this pass, another approach would be to add custom intrinsics within AMDGPU.td. These can be used to create conditions within the ROCDL that lower the required instructions based on the chipset being utilized.

_Generic Solution with Readlane:_
A third potential solution is to attempt a more generalized approach by using readlane and modifying the logic to try to utilize as many common instructions supported across various architectures as possible. This would help in achieving a correct and optimal solution.

stefankoncarevic committed:
Add support for 'row_share' DPP instruction

- Implemented 'row_share' as a new DPP instruction.
- Added verification logic for 'row_share' with permissible
  range [0-15].
- Updated test cases to include 'row_share' examples and checks.

stefankoncarevic committed:
Add ROCDL intrinsic operations for inactive lanes and strict WWM

- Introduced ROCDL_SetInactiveOp to handle inactive lanes by
  mapping to the `llvm.amdgcn.set.inactive` intrinsic. This is
  crucial when the number of active lanes is less, ensuring
  values are managed appropriately.

- Added ROCDL_StrictWWMOp to implement strict whole wavefront mode
  using the `llvm.amdgcn.strict.wwm` intrinsic. This operation
  ensures calculations consider the entire wavefront, integral
  to DPP reduction scenarios.

These additions support complex wavefront operations needed during
DPP (Data Parallel Primitives) reduction, enhancing control over
lane execution and ensuring computation correctness in varying
lane activeness scenarios.

stefankoncarevic committed:
Add PermLaneX16 intrinsic support for RDNA architecture

- Implemented ROCDL_PermLaneX16Op to utilize
  `llvm.amdgcn.permlanex16` intrinsic, addressing limitations in
  DPP broadcasting on RDNA architecture.

- This enhancement facilitates value propagation across lanes
  within a wave, providing flexibility where certain DPP
  instructions are unsupported for broadcasting on RDNA systems.

stefankoncarevic committed:
Implement DPP Reduction in wavefront

- Integrated DPP reduction functionality designed to improve
  performance on supported architectures by minimizing active
  lane usage through set inactive operations.
- Added operations to manage scalar extraction from vectors with
  `vector::ExtractElementOp` for precise per-element manipulation
  and enhanced control over lane values.
- Introduced `ROCDL::SetInactiveOp` to replace values in inactive
  lanes, promoting efficiency when active lane count is reduced
  in DPP scenarios.
- Enabled thorough reduction steps through sequential DPP
  operations including `row_shr`, broadcasting
  (`row_bcast_15`, `row_bcast_31`), and wave rotations
  (`wave_ror`) for wavefront manipulation.
- Utilized `ROCDL::StrictWWMOp` to ensure computations occur in
  Whole Wavefront Mode, particularly crucial in scenarios where
  the number of active lanes is fewer than the wave size. This
  guarantees consistent and correct results across all lanes
  within the wavefront, meeting the necessary computational
  requirements for comprehensive DPP reductions.

stefankoncarevic committed:
Refactor DPP-based broadcast to use ReadlaneOp

Replaced the previous DPP-based broadcasting logic with
ROCDL::ReadlaneOp, allowing thread 0 to extract the final
reduction result and share it with all threads efficiently.
This simplifies the implementation and ensures correctness
across different architectures.

Removed multiple amdgpu::DPPOp operations
(row_share, row_bcast_15, row_bcast_31) used for broadcasting
the reduced value.
Directly used ROCDL::ReadlaneOp to read the result from
thread 0 and distribute it to all threads.
Ensures correct behavior across different AMD GPU architectures.
Potentially improves performance by using a single lane-read
operation instead of multiple DPP-based broadcasts.

stefankoncarevic committed:
Merge branch 'develop' into wave-reduction-dpp

krzysz00 commented:
Happened to notice this PR and wanted to let y'all know that https://github.com/llvm/llvm-project/pull/133204 is a thing that's under development and might be useful.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -1098,31 +1098,132 @@ struct BlockwiseReduceRewritePattern
-                  loc, reduced, workspaceLDSBuffer,
-                  reductionLoop.getLowerCoords(/*domain=*/2));
+
+            Value BrodcastAll;
```
I think we don't need to have rocdl and amdgpu dialects here. Can we have some new op rock::wavereduction or something like that? Then, we can lower it to rocdl later.

dhernandez0 replied:
also, that way it's easier to keep the current implementation if dpp is not supported.

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -1098,31 +1098,132 @@ struct BlockwiseReduceRewritePattern
+                    rewriter.create<arith::ConstantIndexOp>(loc, i));
+                Value scalarInactiveValue = rewriter.create<arith::ConstantOp>(
+                    loc, vecType.getElementType(),
+                    rewriter.getFloatAttr(vecType.getElementType(), 0.0));
```
for max reduction it should be -inf?

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -1098,31 +1098,132 @@ struct BlockwiseReduceRewritePattern
+                    loc, vecType.getElementType(), scalarVal,
+                    scalarInactiveValue);
+
+                Value dppResult1 = rewriter.create<amdgpu::DPPOp>(
```
please add comments to understand what the constants are doing here: 0xF, ...

dhernandez0 commented on mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:
```diff
@@ -1098,31 +1098,132 @@ struct BlockwiseReduceRewritePattern
+                Value dppResult = createReducingOp(op, setInactiveScalar,
+                                                   dppResult1, rewriter);
+
+                Value dppResult2 = rewriter.create<amdgpu::DPPOp>(
```
please add comments to understand what the constants are doing here: 0xF, ...

stefankoncarevic committed:
Add wave_reduce lowering pass and WaveReductionOp
- Introduced `rock-wave-reduce-lowering` pass to
  lower `rock.wave_reduce` ops.
- Implemented `RockWaveReduceRewritePattern` using DPP
  row_shr shifts and broadcasts for intra-wavefront reductions,
  with fallback to `permlanex16` for gfx10+.
- Extended `rocmlir-driver` to pass the detected arch to kernel
  pipeline so the lowering can apply chipset-specific logic.

stefankoncarevic committed:
Updating tests.

stefankoncarevic commented:
This is deprecated, new PR open: https://github.com/ROCm/rocMLIR/pull/2301
